api/ModelHandler.py

```diff
@@ -0,0 +1,314 @@
+
+    if len(minError.index) > 1:
+        # no need to generate new models if error is 0
+        if minError.iloc[0]['errors'] == 0:
+            return
+        else:
+            first = minError.iloc[0]
+            last = minError.iloc[-1]
+
+            biggerFp = first['fp'] > last['fp']
+            smallerFn = first['fn'] < last['fn']
+
+            # Sanity check for bad labels, if the minimum is still the same values
+            # With little labels this could not generate new models
+            if biggerFp or smallerFn:
+                minPenalty = first['penalty']
+                maxPenalty = last['penalty']
+                submitGridSearch(problem, data, minPenalty, maxPenalty)
+
+            return
+
+    else:
+        index = minError.index[0]
+
+        model = minError.iloc[0]
+
+        if model['fp'] > model['fn']:
+            try:
+                above = df.iloc[index + 1]
+            except IndexError:
+                submitOOMJob(problem, data, model['penalty'], '*')
+                return
+
+            minPenalty = model['penalty']
+
+            maxPenalty = above['penalty']
+
+            # If the next model only has 1 more peak, not worth searching
+            if model['numPeaks'] + 1 >= above['numPeaks']:
+                return
+        else:
+            try:
+                below = df.iloc[index - 1]
+            except IndexError:
+                submitOOMJob(problem, data, model['penalty'], '/')
+                return
+
+            minPenalty = below['penalty']
+
+            maxPenalty = model['penalty']
+
+            # If the previous model is only 1 peak away, not worth searching
+            if below['numPeaks'] + 1 >= model['numPeaks']:
+                return
+
+        submitGridSearch(problem, data, minPenalty, maxPenalty)
+
+        return
+
+
+def submitOOMJob(problem, data, penalty, type):
+    job = {'type': 'model', 'problem': problem, 'data': data}
+
+    if type == '*':
+        job['penalty'] = float(penalty) * 10
+    elif type == '/':
+        job['penalty'] = float(penalty) / 10
+    else:
+        print("Invalid OOM Job")
+        return
+    jh.addJob(job)
+
+
+def submitPregenJob(problem, data):
+    job = {'type': 'pregen', 'problem': problem, 'data': data, 'penalties': getPrePenalties(problem, data)}
+    jh.addJob(job)
+
+
+def submitGridSearch(problem, data, minPenalty, maxPenalty, num=pl.gridSearchSize):
+    job = {'type': 'gridSearch', 'problem': problem, 'data': data,
+           'minPenalty': float(minPenalty), 'maxPenalty': float(maxPenalty), 'numModels': num}
+
+    jh.addJob(job)
+
+
+def putModel(data):
+    modelData = data['modelData']
+    modelInfo = data['modelInfo']
+    problem = modelInfo['problem']
+    trackInfo = modelInfo['data']
+    penalty = data['penalty']
+
+    trackPath = '%s%s/%s/' % (pl.dataPath, trackInfo['hub'], trackInfo['track'])
+    problemPath = '%s%s-%d-%d/' % (trackPath, problem['ref'], problem['start'], problem['end'])
```

Comment:
again isn't this logic repetitive with code above?

Reply:
Non factor with DB Implementation